File: Sandbox_Topology/TopologyShared.vb
Imports Rhino.Geometry

Module TopologyShared

    Private Function containsPoint(ByVal _points As List(Of PointTopological), ByVal _check As Point3d, ByVal _T As Double) As Boolean

        For Each _item As PointTopological In _points
            If _item.Point.DistanceTo(_check) < _T Then
                'consider it the same point
                Return True
            End If
        Next

        Return False

    End Function

    Public Function getPointTopo(ByVal P As List(Of Polyline), ByVal _T As Double) As List(Of PointTopological)

        Dim _ptList As New List(Of PointTopological)

        Dim _count As Int32 = 0
        For Each _poly As Polyline In P

            Dim _points As Point3d() = _poly.ToArray

            For i As Int32 = 0 To UBound(_points)

                ' check if point exists in _ptList already
                If Not containsPoint(_ptList, _points(i), _T) Then
                    _ptList.Add(New PointTopological(_points(i), _count))
                    _count = _count + 1
                End If

            Next

        Next

        Return _ptList

    End Function

    Public Function getPLineTopo(ByVal P As List(Of Polyline), ByVal _ptDict As List(Of PointTopological), ByVal _T As Double) As List(Of PLineTopological)

        Dim _lDict As New List(Of PLineTopological)

        Dim _count As Int32 = 0
        For Each _poly As Polyline In P

            Dim _points As Point3d() = _poly.ToArray

            Dim _indices As New List(Of Int32)

            For i As Int32 = 0 To UBound(_points)

                For Each _item As PointTopological In _ptDict
                    If _item.Point.DistanceTo(_points(i)) < _T Then
                        _indices.Add(_item.Index)
                        Exit For
                    End If
                Next

            Next

            _lDict.Add(New PLineTopological(_indices, _count))

            _count = _count + 1

        Next

        Return _lDict

    End Function

    Public Sub setPointPLineTopo(ByVal _lineList As List(Of PLineTopological), ByVal _pointList As List(Of PointTopological))

        For Each _pt As PointTopological In _pointList

            Dim _lList As New List(Of PLineTopological)

            For Each _l As PLineTopological In _lineList

                For Each _index As Int32 In _l.PointIndices
                    If _index = _pt.Index Then
                        _lList.Add(_l)
                        Exit For
                    End If
                Next

            Next

            _pt.PLines = _lList

        Next

    End Sub
End Module

File: Sandbox_Topology/TopologyShared.cs
using System.Collections.Generic;
using Microsoft.VisualBasic;
using Rhino.Geometry;

namespace Sandbox
{

    static class TopologyShared
    {

        private static bool ContainsPoint(List<PointTopological> _points, Point3d _check, double _T)
        {

            foreach (PointTopological _item in _points)
            {
                if (_item.Point.DistanceTo(_check) < _T)
                {
                    // consider it the same point
                    return true;
                }
            }

            return false;

        }

        public static List<PointTopological> GetPointTopo(List<Polyline> P, double _T)
        {

            var _ptList = new List<PointTopological>();

            int _count = 0;
            foreach (Polyline _poly in P)
            {

                var _points = _poly.ToArray();

                for (int i = 0; i < _points.Length; i++)
                {
                    // check if point exists in _ptList already
                    if (!ContainsPoint(_ptList, _points[i], _T))
                    {
                        _ptList.Add(new PointTopological(_points[i], _count));
                        _count += 1;
                    }
                }
            }

            return _ptList;

        }

        public static List<PLineTopological> GetPLineTopo(List<Polyline> P, List<PointTopological> _ptDict, double _T)
        {

            var _lDict = new List<PLineTopological>();

            int _count = 0;
            foreach (Polyline _poly in P)
            {

                var _points = _poly.ToArray();

                var _indices = new List<int>();

                for (int i = 0; i < _points.Length; i++)
                {
                    foreach (PointTopological _item in _ptDict)
                    {
                        if (_item.Point.DistanceTo(_points[i]) < _T)
                        {
                            _indices.Add(_item.Index);
                            break;
                        }
                    }
                }

                _lDict.Add(new PLineTopological(_indices, _count));

                _count += 1;

            }

            return _lDict;

        }

        public static void SetPointPLineTopo(List<PLineTopological> _lineList, List<PointTopological> _pointList)
        {

            foreach (PointTopological _pt in _pointList)
            {

                var _lList = new List<PLineTopological>();

                foreach (PLineTopological _l in _lineList)
                {

                    foreach (int _index in _l.PointIndices)
                    {
                        if (_index == _pt.Index)
                        {
                            _lList.Add(_l);
                            break;
                        }
                    }

                }

                _pt.PLines = _lList;

            }

        }
    }
}